vpr/src/base/noc_storage.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * The NocStorage defines the embedded NoC in the FPGA device.
+ * The NocStorage contains the following information:
+ * - A list of all the routers in the NoC (NocRouter objects)
+ * - A list of all the links in the NoC (NocLink objects)
+ * - A connections list for each router. This is simply a list
+ * of outgoing links of each router.   
+ * There are also a number of functions defined here that can be used
+ * to access the above information or modify them.
+ */
+
+#include "noc_storage.h"
+
+// clear all the graph data structures
+NocStorage::NocStorage() {
+    clear_noc();
+}
+
+// getters for the NoC
+
+// get the outgoing links for a router in the NoC
+const std::vector<NocLinkId>& NocStorage::get_noc_router_connections(NocRouterId id) const {
+    return router_link_list[id];
+}
+
+// get the list of all routers in the NoC
+const vtr::vector<NocRouterId, NocRouter>& NocStorage::get_noc_routers(void) const {
+    return router_storage;
+}
+
+// get the list of all links in the NoC
+const vtr::vector<NocLinkId, NocLink>& NocStorage::get_noc_links(void) const {
+    return link_storage;
+}
+
+// get router properties
+int NocStorage::get_noc_router_grid_position_x(NocRouterId id) const {
+    return router_storage[id].get_router_grid_position_x();
+}
+
+int NocStorage::get_noc_router_grid_position_y(NocRouterId id) const {
+    return router_storage[id].get_router_grid_position_y();
+}
+
+int NocStorage::get_noc_router_id(NocRouterId id) const {
+    return router_storage[id].get_router_id();
+}
+
+std::string NocStorage::get_noc_router_design_module_ref(NocRouterId id) const {
+    return router_storage[id].get_router_design_module_ref();
+}
+
+// get link properties
+NocRouterId NocStorage::get_noc_link_source_router(NocLinkId id) const {
+    return link_storage[id].get_source_router();
+}
+
+NocRouterId NocStorage::get_noc_link_sink_router(NocLinkId id) const {
+    return link_storage[id].get_sink_router();
+}
+
+double NocStorage::get_noc_link_bandwidth_usage(NocLinkId id) const {
```

Comment:
Suggest returning references or const references (your call) to the noc_links and noc_routers so the end caller can just call their functions directly. Otherwise you have to wrap each method in the lower level object with one here in this composite object, which is likely to get tedious.

Reply:
Fixed